### PR TITLE
PP-12354 using service id instead of gateway

### PR DESCRIPTION
**Pull request by james-peacock-ho**
- Add v1/frontend endpoint to get card types for a gateway account by serviceId and accountType (test|live)
- Add v1/api endpoint to get gateway account by serviceId and accountType
- Add unit tests
- Add integration tests and removed some redundant integration tests

NB: where multiple test gateway accounts exist for a service, the one most recently added to the database will be returned. This logic will need refinement in a future story.

### CODE REVIEW

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java [hunk 1]
```diff
@@ -90,6 +112,18 @@ void getAccountShouldNotReturnCardTypes() {
...
+    @Test
+    void getAccountByServiceIdAndAccountTypeShouldNotReturnCredentialsOrCardTypes() {
+        app.createAGatewayAccountWithServiceId("a-valid-service-id");
+        String url = ACCOUNTS_API_SERVICE_ID_URL.replace("{serviceId}", "a-valid-service-id").replace("{accountType}", GatewayAccountType.TEST.name());
+        app.givenSetup()
+                .get(url)
+                .then()
+                .statusCode(200)
+                .body("card_types", is(nullValue()))
+                .body("credentials", is(nullValue()));
```
Comment (DomBelcher) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java]: This is more a problem with the obfuscation of creating the gateway accounts, but it's not clear whether the card_types and credentials are null because they're not returned, or they just don't exist on the account - I think it's actually both in this case, there are no credentials created on the gateway account, but if there were they wouldn't be returned

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java [hunk 2]
```diff
@@ -133,6 +168,20 @@ void getAccountShouldReturn3dsSetting() {
...
+        String gatewayAccountId = app.createAGatewayAccountFor("stripe", "desc", "id", "true", "test");
+        app.getDatabaseTestHelper().updateServiceIdFor(Long.parseLong(gatewayAccountId), "a-valid-service-id");
+        String url = ACCOUNTS_API_SERVICE_ID_URL.replace("{serviceId}", "a-valid-service-id").replace("{accountType}", GatewayAccountType.TEST.name());
+        app.givenSetup()
+                .get(url)
+                .then()
+                .statusCode(200)
+                .body("analytics_id", is("id"))
+                .body("description", is("desc"))
+                .body("requires3ds", is(true));
```
Comment (DomBelcher) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java]: I think this test could be combined with `getAccountByServiceIdAndAccountTypeShouldNotReturnCredentialsOrCardTypes`

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java [hunk 3]
```diff
@@ -168,6 +217,30 @@ void getAccountShouldReturnCorporatePrepaidDebitCardSurchargeAmount() {
...
+                .insert();
+
+        String url = ACCOUNTS_API_SERVICE_ID_URL.replace("{serviceId}", "a-valid-service-id").replace("{accountType}", GatewayAccountType.TEST.name());
+        app.givenSetup()
+                .get(url)
+                .then()
+                .statusCode(200)
+                .body("corporate_credit_card_surcharge_amount", is(corporateCreditCardSurchargeAmount))
+                .body("corporate_debit_card_surcharge_amount", is(corporateDebitCardSurchargeAmount))
+                .body("corporate_prepaid_debit_card_surcharge_amount", is(corporatePrepaidDebitCardSurchargeAmount));
```
Comment (DomBelcher) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java]: This could also be combined with `getAccountByServiceIdAndAccountTypeShouldNotReturnCredentialsOrCardTypes`

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java (hunk 1 above)
Comment (DomBelcher) [reply src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java]: I've dug into this and it seems that this is an outdated test. The `credentials` field on `gateway_accounts` was dropped (#3117) but this test persisted. I assume it can safely be deleted

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceIT.java [hunk 4]
```diff
@@ -136,6 +138,28 @@ void shouldAcceptAllCardTypesNotRequiring3DSForNewlyCreatedAccountAs3dsIsDisable
         validateNon3dsCardType(response, "unionpay", "Union Pay", "CREDIT");
     }
 
+    @Test
+    void shouldGetCardTypesByServiceIdAndAccountType() {
+        app.createAGatewayAccountWithServiceId("a-service-id");
```
Comment (DomBelcher) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceIT.java]: My feeling is we shouldn't use this method of creating accounts, as it works by doing a POST request to the `/v1/api/accounts` endpoint, rather than inserting a record directly into the database, which makes the test depend on that endpoint functioning correctly as well.

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java [hunk 5]
```diff
@@ -269,6 +342,109 @@ void shouldReturnAccountInformationForGetAccountById_withWorldpayCredentials() {
                 .body("gateway_account_credentials[0].credentials.recurring_merchant_initiated", not(hasKey("password")));
     }
 
+    @Test
+    void shouldReturnAccountInformationForGetAccountByServiceIdAndAccountType_withWorldpayCredentials() {
```
Comment (DomBelcher) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java]: The naming of the tests is a bit inconsistent, and they could be broken up a bit to increase readability - e.g. 
`getAccountByServiceIdAndAccountType_withWorldpayCredentials_shouldReturnAccountInformation`

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java [hunk 6]
```diff
@@ -269,6 +342,109 @@ void shouldReturnAccountInformationForGetAccountById_withWorldpayCredentials() {
...
+                                CREDENTIALS_USERNAME, "cit-username",
+                                CREDENTIALS_PASSWORD, "cit-password"),
+                        RECURRING_MERCHANT_INITIATED, Map.of(
+                                CREDENTIALS_MERCHANT_CODE, "mit-merchant-code",
+                                CREDENTIALS_USERNAME, "mit-username",
+                                CREDENTIALS_PASSWORD, "mit-password")))
+                .build();
+
+        this.defaultTestAccount = DatabaseFixtures
+                .withDatabaseTestHelper(app.getDatabaseTestHelper())
```
Comment (DomBelcher) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java:400]: I think this method of creating a test account is better as it works directly with the database. It is also possible to use `app.getDatabaseFixtures()` instead of using the static `DatabaseFixtures`

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java (hunk 5 above)
Comment (DomBelcher) [reply src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java]: Another option would be to use the `@Nested` annotation in JUnit5 to create groups of tests for each method, e.g. having 

```
@Nested
class GetAccountByServiceIdAndAccountType {
  @Test
  void withWorldpayCredentials_shouldReturnAccountInformation() {
    // test body
  }
  // rest of tests...
}
```

which would make it easier to separate the tests for different methods. Could also use the `@DisplayName` annoation to show the API endpoint instead of the class name, e.g. `@DisplayName("GET /v1/services/{serviceId}/{accountType}/account")`

This would be a break from current convention but could be a nice convention to establish going forward - maybe worth further discussion

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java [hunk 7]
```diff
@@ -349,6 +558,18 @@ void shouldNotReturn3dsFlexCredentials_whenGatewayAccountHasNoCreds() {
                 .body("worldpay_3ds_flex", nullValue());
     }
 
+    @Test
+    void shouldNotReturn3dsFlexCredentials_whenGatewayAccountHasNoCreds_byServiceIdAndAccountType_byServiceIdAndAccountType() {
```
Comment (DomBelcher) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java]: Method name duplicates `byServiceIdAndAccountType`

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java [hunk 8]
```diff
@@ -749,6 +970,25 @@ void shouldReturn3dsFlexCredentials_whenGatewayAccountHasCreds() {
                 .body("worldpay_3ds_flex.exemption_engine_enabled", is(false));
     }
 
+    @Test
+    void shouldReturn3dsFlexCredentials_whenGatewayAccountHasCreds_byServiceIdAndAccountType() {
```
Comment (DomBelcher) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java:956]: Could be combined with `shouldReturnAccountInformationForGetAccountByServiceIdAndAccountType_withWorldpayCredentials` - I think we should look to combine the happy path tests into a few which test important cases, e.g. `worldpay` and `stripe` payment providers, and possibly `test` and `live` account types.

**Review by DomBelcher — DISMISSED** [review]
A few things worth discussing/thinking about further

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java (hunk 1 above)
Comment (DomBelcher) [reply src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java]: The `card_types` parameter also appears to be outdated. The `getCardTypes()` method on `GatewayAccountEntity` was previously annotated with `@JsonProperty(card_types)`, so this tested that wasn't returned, but this has since changed. Not sure it's worth keeping around, it certainly feels overkill to have a test dedicated to it

File: src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java [hunk 9]
```diff
@@ -210,6 +227,36 @@ public Response getGatewayAccountAcceptedCardTypes(@Parameter(example = "1", des
...
+                                    "    ]" +
+                                    "}"))),
+                    @ApiResponse(responseCode = "404", description = "Not found")
+            }
+    )
+    public Response getAcceptedCardTypesByServiceIdAndAccountType(@PathParam("serviceId") String serviceId, @PathParam("accountType") GatewayAccountType accountType) {
+        logger.info("Getting accepted card types for service id {}, account type {}", serviceId, accountType.toString());
+        System.out.println("Account type: " + accountType);
+        return gatewayAccountService.getGatewayAccountByServiceIdAndAccountType(serviceId, accountType)
+                .map(gatewayAccount -> successResponseWithEntity(ImmutableMap.of(CARD_TYPES_FIELD_NAME, gatewayAccount.getCardTypes())))
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/gatewayaccount/resource/GatewayAccountResource.java]: Could use the standard Java Map instead of ImmutableMap here...

File: src/test/java/uk/gov/pay/connector/it/dao/GatewayAccountDaoIT.java [hunk 10]
```diff
@@ -646,6 +647,23 @@ void findByExternalId_shouldFindGatewayAccount() {
...
+    void findByServiceIdAndAccountType_shouldFindGatewayAccount() {
+        Long id = nextLong();
+        String externalId = randomUuid();
+        String serviceExternalId = randomUuid();
+        databaseFixtures
+                .aTestAccount()
+                .withAccountId(id)
+                .withExternalId(externalId)
+                .withServiceId(serviceExternalId)
+                .insert();
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/dao/GatewayAccountDaoIT.java]: Can we add another gateway account to make sure the code won't fall over if there are two accounts?

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java [hunk 11]
```diff
@@ -55,41 +56,198 @@ public class GatewayAccountResourceIT {
...
-                .get(ACCOUNTS_API_URL + unknownAccountId)
+                .get(url)
                 .then()
                 .statusCode(404);
     }
 
     @Test
-    void getAccountShouldNotReturnCredentials() {
-        String gatewayAccountId = testBaseExtension.createAGatewayAccountFor("worldpay");
+    void shouldReturnAccountInformation_withWorldpayCredentials_whenGetByServiceIdAndAccountType() {
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java]: Could shorten this to shouldReturnWorldpayAccountInformation_whenGetByServiceIdAndAccountType. The test tests all the account information exists, not just the worldpay credentials so we can remove this from the test name.

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java [hunk 12]
```diff
@@ -55,41 +56,198 @@ public class GatewayAccountResourceIT {
...
+                .body("allow_authorisation_api", is(true))
+                .body("recurring_enabled", is(true))
+                .body("requires3ds", is(true))
+                .body("block_prepaid_cards", is(true))
+                .body("disabled", is(true))
+                .body("disabled_reason", is("Disabled because reasons"))
+                .body("gateway_account_credentials.size()", is(1))
+                .body("gateway_account_credentials[0].payment_provider", is("worldpay"))
+                .body("gateway_account_credentials[0].state", is("ACTIVE"))
+                .body("gateway_account_credentials[0].gateway_account_id", is(accountIdAsInt))
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java]: `accountIdAsInt` is only used here, so it could inlined as `Math.toIntExact(accountId)` and line 135 can go.

File: src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java [hunk 13]
```diff
@@ -55,41 +56,198 @@ public class GatewayAccountResourceIT {
...
+                .body("worldpay_3ds_flex", not(hasKey("jwt_mac_key")))
+                .body("worldpay_3ds_flex", not(hasKey("version")))
+                .body("worldpay_3ds_flex", not(hasKey("gateway_account_id")))
+                .body("worldpay_3ds_flex.exemption_engine_enabled", is(false));
     }
 
     @Test
-    void getAccountShouldNotReturnCardTypes() {
-        String gatewayAccountId = testBaseExtension.createAGatewayAccountFor("worldpay");
+    void shouldReturnAccountInformation_withStripeCredentials_whenGetByServiceIdAndAccountType() {
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceIT.java]: Like the test above, do we want to test for all the returned parameters, rather than just the credentials?

**Review by oswaldquek — CHANGES_REQUESTED** [review]
For the future, a PR for one endpoint would be better :)

**Review by oswaldquek — APPROVED** [review]
